File: webapp/src/modules/audit-management/workflow/auditWorkflowPolicy.ts
/**
 * Single source of truth for audit status transitions and allowed actions.
 * Mirrors AuditWorkflowPolicy.cs — keep both in sync.
 */

export const AuditStatus = {
    Draft:       'Draft',
    Submitted:   'Submitted',
    Reopened:    'Reopened',
    UnderReview: 'UnderReview',
    Approved:    'Approved',
    Distributed: 'Distributed',
    Closed:      'Closed',
} as const;

export type AuditStatusValue = typeof AuditStatus[keyof typeof AuditStatus];

export const CaStatus = {
    Open:       'Open',
    InProgress: 'InProgress',
    Submitted:  'Submitted',
    Closed:     'Closed',
    Voided:     'Voided',
    Overdue:    'Overdue',
} as const;

export type CaStatusValue = typeof CaStatus[keyof typeof CaStatus];

export const AuditWorkflowPolicy = {
    // Edit windows
    auditorEditableStatuses:  [AuditStatus.Draft, AuditStatus.Reopened] as const,
    reviewerEditableStatuses: [AuditStatus.Draft, AuditStatus.Reopened, AuditStatus.UnderReview] as const,

    // Workflow transition gates
    approvableStatuses:   [AuditStatus.UnderReview] as const,
    distributableStatuses:[AuditStatus.Approved, AuditStatus.Distributed] as const,
    reopenableStatuses:   [AuditStatus.Submitted, AuditStatus.UnderReview, AuditStatus.Approved, AuditStatus.Distributed, AuditStatus.Closed] as const,
    closableStatuses:     [AuditStatus.Approved, AuditStatus.Distributed] as const, // UI gate (backend allows more)
    reviewSummaryEditableStatuses: [AuditStatus.UnderReview, AuditStatus.Approved] as const,

    // Statuses where the form is locked (non-reviewer users)
    lockedStatuses: [AuditStatus.Submitted, AuditStatus.UnderReview, AuditStatus.Approved, AuditStatus.Distributed, AuditStatus.Closed] as const,

    // CA status helpers
    caOpenStatuses:     [CaStatus.Open, CaStatus.InProgress, CaStatus.Submitted, CaStatus.Overdue] as const,
    caTerminalStatuses: [CaStatus.Closed, CaStatus.Voided] as const,
} as const;

/** Severity tag for a given audit status (PrimeVue Tag severity). */
export function auditStatusSeverity(status: string): string {
    const map: Record<string, string> = {
        Draft:       'warning',
        Reopened:    'warning',
        Submitted:   'info',
        UnderReview: 'info',
        Approved:    'success',
        Distributed: 'success',
        Closed:      'secondary',
    };
    return map[status] ?? 'secondary';
}

/** Severity for a CA status badge. */
export function caStatusSeverity(status: string): string {
    const map: Record<string, string> = {
        Open:       'danger',
        InProgress: 'warning',
        Submitted:  'info',
        Overdue:    'danger',
        Closed:     'success',
        Voided:     'secondary',
    };
    return map[status] ?? 'secondary';
}

File: Api/Domain/Audit/AuditWorkflowPolicy.cs
namespace Stronghold.AppDashboard.Api.Domain.Audit;

/// <summary>
/// Single source of truth for audit status transitions and allowed actions.
/// Use these constants instead of scattering string arrays across handlers and controllers.
/// </summary>
public static class AuditWorkflowPolicy
{
    // ── Status constants ──────────────────────────────────────────────────────

    public const string Draft       = "Draft";
    public const string Submitted   = "Submitted";
    public const string Reopened    = "Reopened";
    public const string UnderReview = "UnderReview";
    public const string Approved    = "Approved";
    public const string Distributed = "Distributed";
    public const string Closed      = "Closed";

    // ── Edit windows ──────────────────────────────────────────────────────────

    /// <summary>Statuses where an auditor-only user may save responses.</summary>
    public static readonly string[] AuditorEditableStatuses = [Draft, Reopened];

    /// <summary>Statuses where an admin/reviewer may save responses.</summary>
    public static readonly string[] ReviewerEditableStatuses = [Draft, Reopened, UnderReview];

    // ── Workflow transition gates ─────────────────────────────────────────────

    /// <summary>Statuses from which StartReview is valid.</summary>
    public static readonly string[] ReviewableStatuses = [Submitted];

    /// <summary>Statuses from which ApproveAudit is valid.</summary>
    public static readonly string[] ApprovableStatuses = [UnderReview];

    /// <summary>Statuses from which SendDistributionEmail is valid.</summary>
    public static readonly string[] DistributableStatuses = [Approved, Distributed];

    /// <summary>Statuses from which ReopenAudit is valid.</summary>
    public static readonly string[] ReopenableStatuses = [Submitted, UnderReview, Approved, Distributed, Closed];

    /// <summary>Statuses from which CloseAudit is valid (only after workflow completes).</summary>
    public static readonly string[] ClosableStatuses = [Approved, Distributed];

    /// <summary>Statuses where SaveReviewSummary is valid.</summary>
    public static readonly string[] ReviewSummaryEditableStatuses = [UnderReview, Approved];

    // ── CA status constants ───────────────────────────────────────────────────

    public const string CaOpen       = "Open";
    public const string CaInProgress = "InProgress";
    public const string CaSubmitted  = "Submitted";
    public const string CaClosed     = "Closed";
    public const string CaVoided     = "Voided";
    public const string CaOverdue    = "Overdue";

    /// <summary>CA statuses that are considered "resolved" (closed-out).</summary>
    public static readonly string[] CaTerminalStatuses = [CaClosed, CaVoided];

    /// <summary>CA statuses that are considered "open" for distribution filtering.</summary>
    public static readonly string[] CaOpenStatuses = [CaOpen, CaInProgress, CaSubmitted, CaOverdue];

    /// <summary>External assignees may set these statuses via public token.</summary>
    public static readonly string[] CaExternalAllowedStatuses = [CaInProgress, CaSubmitted];
}
